From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Sat, 28 Jan 2023 14:10:43 -0500
Subject: [PATCH] Dont close Slime Plugin Classloader


diff --git a/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java b/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
index 7572a0bf6614b02be3cbccc7b86e52ee1b8df621..c387e39a332a957dcf2647bba3a2e31b7b6c63b5 100644
--- a/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
+++ b/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
@@ -256,7 +256,9 @@ public final class PluginClassLoader extends URLClassLoader implements io.paperm
             super.close();
         } finally {
             this.closed = true; // Pufferfish
+            if (this.plugin == null || !this.plugin.getName().equals("SlimeWorldManager")) { // ASWM - Don't close
             jar.close();
+            } // ASWM - Don't close
         }
     }
 
