From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Sat, 28 Jan 2023 14:10:43 -0500
Subject: [PATCH] Dont close Slime Plugin Classloader


diff --git a/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java b/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
index 90953bfc81168068a281be4d2d3942d5e7dd69ff..24b6fb94c88d4d544eed25c4199d59f003f2e5b8 100644
--- a/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
+++ b/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
@@ -270,7 +270,9 @@ public final class PluginClassLoader extends URLClassLoader implements io.paperm
             super.close();
         } finally {
             this.closed = true; // Pufferfish
+            if (this.plugin == null || !this.plugin.getName().equals("SlimeWorldManager")) { // ASWM - Don't close
             jar.close();
+            } // ASWM - Don't close
         }
     }
 
